src/linear_algebra/orientation.lean

```diff
@@ -0,0 +1,376 @@
+lemma same_orientation_pos_smul_right (f : alternating_map R M R ι) {r : R} (h : 0 < r) :
+  same_orientation f (r • f) :=
+⟨r, 1, h, zero_lt_one, (one_smul _ _).symm⟩
+
+/-- A positive multiple of an alternating map has the same orientation as that map. -/
+lemma same_orientation_pos_smul_left (f : alternating_map R M R ι) {r : R} (h : 0 < r) :
+  same_orientation (r • f) f :=
+⟨1, r, zero_lt_one, h, one_smul _ _⟩
+
+end nontrivial
+
+variables {R M ι}
+
+/-- If two alternating maps have the same orientation, so do their negations. -/
+@[simp] lemma same_orientation_neg_iff (f g : alternating_map R M R ι) :
+  same_orientation (-f) (-g) ↔ same_orientation f g :=
+⟨λ ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₁, r₂, hr₁, hr₂, by rwa [smul_neg, smul_neg, neg_inj] at h⟩,
+ λ ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₁, r₂, hr₁, hr₂, by rwa [smul_neg, smul_neg, neg_inj]⟩⟩
+
+end alternating_map
+
+variables (R M ι)
+
+/-- Nonzero alternating maps, as used to define orientations. -/
+@[reducible] def orientation_map := {f : alternating_map R M R ι // f ≠ 0}
+
+instance : has_coe_to_fun (orientation_map R M ι) (λ _, (ι → M) → R) :=
+⟨λ x, ⇑(x : alternating_map R M R ι)⟩
+
+namespace orientation_map
+
+variables {R M ι}
+
+/-- Negating a nonzero alternating map. -/
+instance : has_neg (orientation_map R M ι) :=
+⟨λ ⟨f, hf⟩, ⟨-f, by rw neg_ne_zero; exact hf⟩⟩
+
+/-- Negating a nonzero alternating map commutes with coercion to `alternating_map`. -/
+@[simp, norm_cast] lemma coe_neg (f : orientation_map R M ι) :
+  ((-f : orientation_map R M ι) : alternating_map R M R ι) = -(f : alternating_map R M R ι) :=
+by cases f; refl
+
+/-- Negating a nonzero alternating map twice produces the original map. -/
+@[simp] protected lemma neg_neg (f : orientation_map R M ι) : -(-f) = f :=
+by rw [subtype.ext_iff, coe_neg, coe_neg, neg_neg]
+
+variables (R M ι)
+
+variables [nontrivial R]
+
+/-- The setoid of the `same_orientation` relation for the subtype of nonzero alternating maps. -/
+instance same_orientation_setoid : setoid (orientation_map R M ι) :=
+{ r := λ f g, (f : alternating_map R M R ι).same_orientation g,
+  iseqv := equivalence.comap (alternating_map.equivalence_same_orientation R M ι) _ }
+
+variables {R M ι}
+
+/-- Equivalence of nonzero alternating maps, in terms of same_orientation. -/
+lemma equiv_iff_same_orientation (f g : orientation_map R M ι) :
+  f ≈ g ↔ (f : alternating_map R M R ι).same_orientation g :=
+iff.rfl
+
+/-- If two nonzero alternating maps are equivalent, so are their negations. -/
+@[simp] lemma equiv_neg_iff (f g : orientation_map R M ι) : -f ≈ -g ↔ f ≈ g :=
+by rw [equiv_iff_same_orientation, equiv_iff_same_orientation, coe_neg, coe_neg,
+       alternating_map.same_orientation_neg_iff]
+
+end orientation_map
+
+variables [nontrivial R]
+
+/-- An orientation of a module, intended to be used when `ι` is a `fintype` with the same
+cardinality as a basis. -/
+@[ext] class orientation :=
+(alt_maps : quotient (orientation_map.same_orientation_setoid R M ι))
```

Comment:
I'm curious what the trade-off is if you instead do something like:
```suggestion
(alt_maps : quotient (alternating_map.same_orientation_setoid R M ι))
(ne_zero : quotient.lift (≠ 0) sorry alt_maps)
```

Reply:
A zero orientation definitely isn't interesting itself; pretty much anything using orientations requires one that's nonzero. So the condition of being nonzero needs to go there somewhere. Then, I think the main trade-off between having the `setoid` on alternating maps versus on a subtype `orientation_map` (or `ray_vector` or similar if everything is done for rays from the origin in a general module rather than just that of alternating maps) is that, on a subtype defined specifically for this purpose, it seems reasonable for the `setoid` to be a (global) instance: the only equivalence relation to be used on that type. Whereas if a `setoid` is used for alternating maps (or a general module), it doesn't seem appropriate for it to be an instance globally (people might want to use other equivalence relations as well), only locally, which is less convenient.

What should it seems be avoided is having a `setoid` on both alternating maps and the subtype. I originally tried that, and ran into diamond issues (`has_equiv` for `orientation_map` being deduced from either the `setoid` on `orientation_map`, or from `subtype.has_equiv`, with those not being defeq).

Reply:
Can you give a mwe for the diamond? I'm surprised it's not defeq.

At any rate, I think the availability of a global setoid instance isn't that important; in my experience we use `quotient.mk'` (no typeclass argument) more often than `quotient.mk`. See for instance quotients by subgroups or ideals, etc.

Reply:
Regarding zero orientations; don't all your lemmas so far apply even when the orientation is not zero?

Reply:
I didn't keep any versions of the code with the diamond. From memory, it used the existing `setoid` instance for subtypes to deduce the `setoid` instance for `orientation_map` from that for `alternating_map` (and the diamond problem was from the two paths to `has_equiv` on the subtype).

Reply:
With changes to work on general modules, the `setoid` instance now needs to be a local instance anyway (because `ray_vector` takes only `M` as a type class argument, but the instance depends on `R` as well), but the actual definition using the quotient is still slightly simpler (using a quotient directly rather than putting it in a structure).